File: src/omnia/data/iterate/IntegerRangeIterator.java
package omnia.data.iterate;

import java.util.Iterator;

public final class IntegerRangeIterator implements Iterator<Integer> {
  private int next;
  private final int end;
  private final int increment;

  public static IntegerRangeIterator create(int start, int end) {
    return new IntegerRangeIterator(start, end, start <= end ? 1 : -1);
  }

  private IntegerRangeIterator(int start, int end, int increment) {
    this.next = start;
    this.end = end;
    this.increment = increment;
  }

  @Override
  public boolean hasNext() {
    return increment > 0
        ? next + increment < end
        : next + increment > end;
  }

  @Override
  public Integer next() {
    int current = next;
    next += increment;
    return current;
  }
}

File: src/omnia/data/iterate/IntegerRangeIterator.kt
package omnia.data.iterate

class IntegerRangeIterator private constructor(
  private var next: Int,
  private val end: Int,
  private val increment: Int
) : Iterator<Int> {

  override fun hasNext(): Boolean {
    return if (increment > 0) next + increment < end else next + increment > end
  }

  override fun next(): Int {
    val current = next
    next += increment
    return current
  }

  companion object {

    fun create(start: Int, end: Int): IntegerRangeIterator {
      return IntegerRangeIterator(start, end, if (start <= end) 1 else -1)
    }
  }
}